rcl/include/rcl/node_type_cache.h

```diff
@@ -0,0 +1,197 @@
+// Copyright 2023 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCL__NODE_TYPE_CACHE_H_
+#define RCL__NODE_TYPE_CACHE_H_
+
+#include "rcl/node.h"
+#include "rcl/types.h"
+#include "rcl/visibility_control.h"
+#include "rosidl_runtime_c/type_description/type_description__struct.h"
+#include "rosidl_runtime_c/type_description/type_source__struct.h"
+#include "type_description_interfaces/msg/type_description.h"
+#include "type_description_interfaces/msg/type_source.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct rcl_type_info_t
+{
+  type_description_interfaces__msg__TypeDescription * type_description;
+  type_description_interfaces__msg__TypeSource__Sequence * type_sources;
+} rcl_type_info_t;
+
+/// Initialize the node's type cache.
+/**
+ * This function initializes hash map of the node's type cache such that types
+ * can be registered and retrieved.
+ * Note that to correctly capture all types used by a node, this needs to be called
+ * before any "builtin" publishers or services are created.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] node the handle to the node whose type cache should be initialized
+ * \return #RCL_RET_OK if the node's type cache was successfully initialized, or
+ * \return #RCL_RET_INVALID_ARGUMENT if any arguments are invalid, or
+ * \return #RCL_RET_NODE_INVALID if the given `node` is invalid, or
+ * \return #RCL_RET_ERROR if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t rcl_node_type_cache_init(rcl_node_t * node);
+
+/// Check whether node's type cache has been initialized.
+/**
+ * Based on configuration, the node may optionally disable the type cache,
+ * this convenience function allows a standardized check for it.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] node the handle to the node whose type cache to check
+ * \return true if the node's type cache was successfully initialized, or
+ * \return false for any other case.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+bool rcl_node_type_cache_is_valid(const rcl_node_t * node);
```

Comment:
Given that we are unconditionally calling `rcl_node_type_cache_init` during `rcl_node_init` now, do we need this function?

Reply:
Nope - removing